crates/solver/src/settlement_submission/submitter/eden_api.rs

```diff
@@ -97,13 +98,21 @@ impl TransactionSubmitting for EdenApi {
                     .submit_raw_transaction(tx)
                     .await
             })
-            .await
+            .await;
+
+        let successful = match &result {
+            Ok(_) => true,
+            // Sometimes `submit_slot_transaction()` times out and the fallback submission
+            // strategy reveals that the network is already aware of the transaction by returning
+            // the error message "already known".
+            Err(err) => err.to_string().contains("already known"),
+        };
+        super::track_submission_success("eden", successful);
```

Comment:
Instead of tracking submission in each submitter implementation, should we instead introduce an `InstrumentedSubmitter` instead? My concern is that if we add a new submitter we may forget to instrument it.

I'm on the fence about this suggestion - just wanted to bring it up.

Reply:
> Eden is the best example here: AFAIK it's the only submitter where we can say a submission was successful even if we are not able to produce a TransactionHandle.

Right, so that would require even more work to change the interface (to bubble up the fact of whether or not an error is a "hard" error or a "soft" error).

This also feels icky because you are modifying the interface in order to introduce a instrumenting component further up the call-chain.

With that in mind, I think it makes sense in this case to instrument each submitter individually then.

Reply:
I'm more in favor of removing the `submit_slot_transaction` call since we are not going to bid for Eden slot anymore. It's unreliable and forcing us to have these kind of weird cases to handle.

Reply:
Good point. That probably makes sense in a separate PR though (since its unrelated to instrumentation right?).

Reply:
So if we consider dropping `submit_slot_transaction()` the `InstrumentedSubmitter` idea seems way nicer but I encountered one problem while implementing it:

At the moment we have a single instance of `CustomNodesApi` which handles any number of nodes. To have metrics for each individual node we would have to have a `CustomNodesApi` instance per node. This is because the `CustomNodesApi` API only gives us the information that either all nodes failed or some node succeeded.

This can be solved by either leaving the code mostly as is and have n `CustomNodesApi` instances each having handling a `Vec` of nodes with size 1. Or we could rewrite `CustomNodesApi` to always handle a single node.

IMO the cleaner approach would be the latter but in the interest of time and small PRs I can do the former and change the `CustomNodesApi` afterwards.

Reply:
Note that if `CustomNodesApi` contains only one node, then we will have 5 transaction strategies (Eden + Flashbots + 3nodes) instead of 3 (Eden + Flashbots + CustomNodes), meaning our submission component will now run this function additional two times for two additional nodes without actual need (all 3 nodes have exact code execution while Eden and Flashbots have some specifics and that was the reason to parallelize the execution at the level of `Submitter` and not at the level of `Submitter::TransactionSubmitting`: https://github.com/gnosis/gp-v2-services/blob/c2abb1955aa7909258264b92dbeb4ceb7d623621/crates/solver/src/settlement_submission/submitter.rs#L211

Reply:
That's unfortunate. :(
The only alternative I see would be to create another type implementing `TransactionSubmitting` (let's say `CustomNode`) which takes only a single node (`Web3`) and make it that `CustomNodesApi` operates on a `Vec<Box<dyn TransactionSubmitting>>` instead of `Vec<Web3>`. That way we could wrap `CustomNode` in an `InstrumentedNode` and pass a vector of those to `CustomNodesApi`.
`CustomNodesApi` would then be similar to the `RacingPriceEstimator` but a little bit simpler.

Reply:
To me the instrumenting part is secondary and it feels like we're having to make the primary logic more complicated to make the instrumenting nicer which I dislike. The concern about forgetting to add instrumenting to a new submitter doesn't outweight these disadvantages.
(But I haven't looked at this code in detail.)